### PR TITLE
BC-295 - improve security (npm ci ignores NODE_ENV=production)

### PR DESCRIPTION
**Pull request by virgilchiriac**
# Short Description

Because npm ci ignores NODE_ENV=production, the build now explicitly needs to ignore dev dependencies so that are not part of the build image.

## Links to Ticket and related Pull-Requests
- https://ticketsystem.hpi-schul-cloud.org/browse/BC-295

## Checklist before merging

- [ ] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [ ] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: Dockerfile.client [hunk 1]
```diff
@@ -7,7 +7,9 @@ WORKDIR '/app'
 
 COPY package.json package-lock.json ./
 USER node
-RUN npm ci
+
+RUN rm -rf node_modules && npm cache clean --force && npm install --production
```
Comment (mamutmk5) [inline Dockerfile.client:11]: I think olny run _npm install --production_ shoud do it.
Why: ist starts with an folder who only inclued to files _package.json_ and _package-lock.json_. There is no folder like node_modules or even something at the npm cahce.
```suggestion
RUN npm install --production
```

File: .dockerignore [hunk 2]
```diff
@@ -1,7 +1,30 @@
...
+.sass-lint-auto-fix.yml
+.sass-lint.yml
+.travis.yml*
 .vscode
+.webpack-changed-plugin-cache
+CHANGELOG.md
 deploy
+Dockerfile*
+README.md
+ansible
```
Comment (mamutmk5) [inline .dockerignore:22]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 3]
```diff
@@ -1,7 +1,30 @@
...
+.idea
+.mergify.yml
 .nuxt
+.pullapprove.yml
+.sass-lint-auto-fix.yml
+.sass-lint.yml
+.travis.yml*
 .vscode
+.webpack-changed-plugin-cache
+CHANGELOG.md
```
Comment (mamutmk5) [inline .dockerignore:18]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 4]
```diff
@@ -1,7 +1,30 @@
+.DS_Store
+.build
+.eslintrc.json
```
Comment (mamutmk5) [inline .dockerignore:3]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 5]
```diff
@@ -1,7 +1,30 @@
...
+.git
+.gitattributes
 .github
+.gitignore
+.gulp-changed-smart.json
+.idea
+.mergify.yml
```
Comment (mamutmk5) [inline .dockerignore:10]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 6]
```diff
@@ -1,7 +1,30 @@
...
 .nuxt
+.pullapprove.yml
```
Comment (mamutmk5) [inline .dockerignore:12]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 7]
```diff
@@ -1,7 +1,30 @@
...
+.sass-lint-auto-fix.yml
+.sass-lint.yml
+.travis.yml*
```
Comment (mamutmk5) [inline .dockerignore:15]: File shoud be not more exist

File: .dockerignore [hunk 8]
```diff
@@ -1,7 +1,30 @@
...
+.webpack-changed-plugin-cache
+CHANGELOG.md
 deploy
+Dockerfile*
+README.md
+ansible
+build
+compose-client-test.dummy
+deploy.sh
+docs
```
Comment (mamutmk5) [inline .dockerignore:26]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 9]
```diff
@@ -1,7 +1,30 @@
...
 node_modules
-.travis.yml*
-Dockerfile.*
+test
```
Comment (mamutmk5) [inline .dockerignore:28]: Must be include at the immage, by AGLPv3

File: .dockerignore [hunk 10]
```diff
@@ -1,7 +1,30 @@
...
+travis*
+webpack.config.js
```
Comment (mamutmk5) [inline .dockerignore:30]: Must be include at the immage, by AGLPv3

**Review by Loki-Afro — CHANGES_REQUESTED** [review]
please wait with this merge until BC-266 is through, you may need to change things

**Review by mamutmk5 — CHANGES_REQUESTED** [review]
Please Respect the **AGPv3** and the _6. Conveying Non-Source Forms_.
That's some how a docker image for me.

File: Dockerfile.client (hunk 1 above)
Comment (Loki-Afro) [reply Dockerfile.client]: doing npm install will ignore the package-lock.json leading to non reproduce able builds - don't do that!

**Review by Loki-Afro — CHANGES_REQUESTED** [review]
@mamutmk5 any source for your AGPLv3 comments? i would rather have the smallest image possible

**Comment by virgilchiriac** [comment]
nuxt build depends on webpack, which is a deep devDependency. 
because we run  nuxt build also on startup, this needs to wait for BC-266

**Review by Loki-Afro — CHANGES_REQUESTED** [review]
do not merge this before bc-266

and use npm ci instead npm install - you want that you verify the package-json.lock,otherwise we have unstable builds.

fix your dev dependencies as well, it is worth nothing if your local system is compromised.

and/or find a way to use npm ci and only install prod dependencies

**Comment by virgilchiriac** [comment]
the build needs dev dependencies, therefor we need to rethink the image artifacts: instead of copy all items, we should only use the ones needed to run the application

**Pull request closed without merging**